File: JoyReactor.Android/Widget/FixedAspectPanel.cs

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Views;
using Android.Widget;
using Android.Util;

namespace JoyReactor.Android.Widget
{
	public class FixedAspectPanel : ViewGroup
	{
		private float _aspect = 1;

		public float Aspect {
			get { return _aspect; }
			set {
				_aspect = value;
				RequestLayout ();
			}
		}

		public FixedAspectPanel (Context context, IAttributeSet attrs) : base (context, attrs)
		{
		}

		protected override void OnMeasure (int widthMeasureSpec, int heightMeasureSpec)
		{
			int w, h;
			if (MeasureSpec.GetSize (widthMeasureSpec) != 0) {
				w = MeasureSpec.GetSize (widthMeasureSpec);
				h = (int)(w / _aspect);
			} else if (MeasureSpec.GetSize (widthMeasureSpec) != 0) {
				h = MeasureSpec.GetSize (heightMeasureSpec);
				w = (int)(h * _aspect);
			} else throw new Exception ();
			SetMeasuredDimension (w, h);

			for (int i = 0; i < ChildCount; i++) {
				GetChildAt (i).Measure (
					MeasureSpec.MakeMeasureSpec (w, MeasureSpecMode.Exactly),
					MeasureSpec.MakeMeasureSpec (h, MeasureSpecMode.Exactly));
			}
		}

		#region implemented abstract members of ViewGroup

		protected override void OnLayout (bool changed, int l, int t, int r, int b)
		{
			for (int i = 0; i < ChildCount; i++) {
				GetChildAt (i).Layout (l, t, r, b);
			}
		}

		#endregion
	}
}

File: android/src/main/kotlin/y2k/joyreactor/widget/FixedAspectPanel.kt
package y2k.joyreactor.widget

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import y2k.joyreactor.common.forEachChild
import y2k.joyreactor.common.measureSpec
import kotlin.properties.Delegates

/**
 * Created by y2k on 9/26/15.
 */
class FixedAspectPanel(context: Context, attrs: AttributeSet) : ViewGroup(context, attrs) {

    var aspect by Delegates.observable(1f) { prop, old, new ->
        if (old != new) requestLayout()
    }

    init {
        if (isInEditMode) aspect = 2f
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val w: Int
        val h: Int
        if (MeasureSpec.getSize(widthMeasureSpec) != 0) {
            w = MeasureSpec.getSize(widthMeasureSpec)
            h = (w / aspect).toInt()
        } else if (MeasureSpec.getSize(heightMeasureSpec) != 0) {
            h = MeasureSpec.getSize(heightMeasureSpec)
            w = (h * aspect).toInt()
        } else {
            throw IllegalStateException()
        }
        setMeasuredDimension(w, h)

        forEachChild {
            it.measure(
                w.measureSpec(MeasureSpec.EXACTLY),
                h.measureSpec(MeasureSpec.EXACTLY))
        }
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        forEachChild { it.layout(0, 0, r - l, b - t) }
    }
}